File: resources/ios/TextFieldControl.swift
import SwiftUI
import UIKit

struct FirstlightTextFieldControl: View {
    let configuration: TextFieldRendererConfiguration
    @Binding var text: String
    @Binding var revealed: Bool
    let isFocused: FocusState<Bool>.Binding
    let tokens: NativeUITokens
    let onClear: () -> Void
    let onTrailingPress: () -> Void
    let onSubmit: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            if !configuration.label.isEmpty {
                Text(configuration.required ? "\(configuration.label) *" : configuration.label)
                .font(.subheadline.weight(.medium))
                .foregroundStyle(tokens.onSurface)
            }

            HStack(spacing: 8) {
                if !configuration.leadingIcon.isEmpty {
                    Image(systemName: configuration.leadingIcon)
                        .foregroundStyle(tokens.onSurfaceVariant)
                        .accessibilityHidden(true)
                }

                input
                    .focused(isFocused)
                    .disabled(configuration.disabled)

                trailing
            }
            .padding(.horizontal, 12)
            .frame(minHeight: 44)
            .background(Color(uiColor: .secondarySystemBackground))
            .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
            .overlay {
                RoundedRectangle(cornerRadius: 10, style: .continuous)
                    .stroke(configuration.error.isEmpty ? Color.clear : tokens.destructive, lineWidth: 1)
            }

            let supporting = configuration.error.isEmpty ? configuration.helper : configuration.error
            if !supporting.isEmpty {
                Text(supporting)
                    .font(.footnote)
                    .foregroundStyle(configuration.error.isEmpty ? tokens.onSurfaceVariant : tokens.destructive)
                    .accessibilityLabel(configuration.error.isEmpty ? supporting : "Error: \(supporting)")
            }
        }
        .opacity(configuration.disabled ? 0.6 : 1)
    }

    @ViewBuilder private var input: some View {
        if configuration.secure && !revealed {
            SecureField(
                configuration.placeholder,
                text: configuration.readOnly ? .constant(text) : $text
            )
                .textContentType(resolveContentType(configuration.contentType))
                .keyboardType(resolveKeyboard(configuration.keyboard))
                .textInputAutocapitalization(resolveCapitalization(configuration.autocapitalize))
                .autocorrectionDisabled(configuration.autocorrectPolicy == "disabled")
                .submitLabel(resolveSubmitLabel(configuration.submitLabel))
                .onSubmit(onSubmit)
        } else {
            TextField(
                configuration.placeholder,
                text: configuration.readOnly ? .constant(text) : $text
            )
                .textContentType(resolveContentType(configuration.contentType))
                .keyboardType(resolveKeyboard(configuration.keyboard))
                .textInputAutocapitalization(resolveCapitalization(configuration.autocapitalize))
                .autocorrectionDisabled(configuration.autocorrectPolicy == "disabled")
                .submitLabel(resolveSubmitLabel(configuration.submitLabel))
                .onSubmit(onSubmit)
        }
    }

    @ViewBuilder private var trailing: some View {
        if configuration.clearable && !text.isEmpty && !configuration.readOnly {
            actionButton(symbol: "xmark.circle.fill", label: configuration.clearA11yLabel, action: onClear)
        } else if configuration.revealable {
            actionButton(
                symbol: revealed ? "eye.slash" : "eye",
                label: revealed ? configuration.hidePasswordA11yLabel : configuration.showPasswordA11yLabel,
                action: { revealed.toggle() }
            )
        } else if !configuration.trailingIcon.isEmpty {
            if configuration.onPressCallback != 0 {
                actionButton(
                    symbol: configuration.trailingIcon,
                    label: configuration.trailingAccessibilityLabel,
                    action: onTrailingPress
                )
            } else {
                Image(systemName: configuration.trailingIcon)
                    .foregroundStyle(tokens.onSurfaceVariant)
                    .accessibilityHidden(true)
            }
        }
    }

    private func actionButton(symbol: String, label: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Image(systemName: symbol).frame(minWidth: 44, minHeight: 44)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(Text(label))
        .disabled(configuration.disabled)
    }
}

private func resolveKeyboard(_ value: String) -> UIKeyboardType {
    switch value {
    case "email": .emailAddress
    case "phone": .phonePad
    case "url": .URL
    case "number": .numberPad
    case "decimal": .decimalPad
    default: .default
    }
}

private func resolveContentType(_ value: String) -> UITextContentType? {
    switch value {
    case "name": .name
    case "username": .username
    case "email": .emailAddress
    case "password": .password
    case "new-password": .newPassword
    case "one-time-code": .oneTimeCode
    default: nil
    }
}

private func resolveCapitalization(_ value: String) -> TextInputAutocapitalization? {
    switch value {
    case "none": .never
    case "sentences": .sentences
    case "words": .words
    case "characters": .characters
    default: nil
    }
}

private func resolveSubmitLabel(_ value: String) -> SubmitLabel {
    switch value {
    case "go": .go
    case "next": .next
    case "search": .search
    case "send": .send
    default: .return
    }
}

File: resources/android/TextFieldControl.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.autofill.ContentType
import androidx.compose.ui.autofill.contentType
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import com.nativephp.mobile.ui.MaterialIcon
import com.nativephp.plugins.native_ui.NativeUITokens

@Composable
internal fun FirstlightTextFieldControl(
    configuration: TextFieldRendererConfiguration,
    draft: TextFieldValue,
    revealed: Boolean,
    tokens: NativeUITokens,
    onValueChange: (TextFieldValue) -> Unit,
    onClear: () -> Unit,
    onReveal: () -> Unit,
    onTrailingPress: () -> Unit,
    onSubmit: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val supporting = configuration.error.ifEmpty { configuration.helper }
    val description = listOf(configuration.accessibilityLabel, configuration.accessibilityHint)
        .filter(String::isNotEmpty)
        .joinToString(". ")

    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        OutlinedTextField(
            value = draft,
            onValueChange = onValueChange,
            modifier = modifier
                .fillMaxWidth()
                .defaultMinSize(minHeight = 48.dp)
                .firstlightContentType(configuration.contentType)
                .semantics {
                    if (description.isNotEmpty()) contentDescription = description
                    if (configuration.error.isNotEmpty()) error(configuration.error)
                },
            enabled = !configuration.disabled,
            readOnly = configuration.readOnly,
            singleLine = true,
            isError = configuration.error.isNotEmpty(),
            label = configuration.label.takeIf(String::isNotEmpty)?.let { label ->
                { Text(if (configuration.required) "$label *" else label) }
            },
            placeholder = configuration.placeholder.takeIf(String::isNotEmpty)?.let { text ->
                { Text(text) }
            },
            supportingText = supporting.takeIf(String::isNotEmpty)?.let { text ->
                { Text(text, color = if (configuration.error.isEmpty()) tokens.onSurfaceVariant else tokens.destructive) }
            },
            leadingIcon = configuration.leadingIcon.takeIf(String::isNotEmpty)?.let { name ->
                { MaterialIcon(name, null) }
            },
            trailingIcon = trailingSlot(
                configuration = configuration,
                hasValue = draft.text.isNotEmpty(),
                revealed = revealed,
                onClear = onClear,
                onReveal = onReveal,
                onTrailingPress = onTrailingPress,
            ),
            visualTransformation = if (configuration.secure && !revealed) {
                PasswordVisualTransformation()
            } else {
                VisualTransformation.None
            },
            keyboardOptions = textFieldKeyboardOptions(configuration),
            keyboardActions = textFieldKeyboardActions(configuration, onSubmit),
        )
    }
}

private fun Modifier.firstlightContentType(value: String): Modifier {
    val type = when (value) {
        "name" -> ContentType.PersonFullName
        "username" -> ContentType.Username
        "email" -> ContentType.EmailAddress
        "password" -> ContentType.Password
        "new-password" -> ContentType.NewPassword
        "one-time-code" -> ContentType.SmsOtpCode
        else -> null
    }

    return type?.let(::contentType) ?: this
}

private fun trailingSlot(
    configuration: TextFieldRendererConfiguration,
    hasValue: Boolean,
    revealed: Boolean,
    onClear: () -> Unit,
    onReveal: () -> Unit,
    onTrailingPress: () -> Unit,
): (@Composable () -> Unit)? = when {
    configuration.clearable && hasValue && !configuration.readOnly -> ({
        textFieldIconButton("close", configuration.clearA11yLabel, configuration.disabled, onClear)
    })
    configuration.revealable -> ({
        textFieldIconButton(
            if (revealed) "visibility_off" else "visibility",
            if (revealed) configuration.hidePasswordA11yLabel else configuration.showPasswordA11yLabel,
            configuration.disabled,
            onReveal,
        )
    })
    configuration.trailingIcon.isNotEmpty() && configuration.onPressCallback != 0 -> ({
        textFieldIconButton(
            configuration.trailingIcon,
            configuration.trailingAccessibilityLabel,
            configuration.disabled,
            onTrailingPress,
        )
    })
    configuration.trailingIcon.isNotEmpty() -> ({ MaterialIcon(configuration.trailingIcon, null) })
    else -> null
}

@Composable
private fun textFieldIconButton(name: String, label: String, disabled: Boolean, action: () -> Unit) {
    IconButton(
        onClick = action,
        enabled = !disabled,
        modifier = Modifier
            .sizeIn(minWidth = 48.dp, minHeight = 48.dp)
            .semantics { contentDescription = label },
    ) {
        MaterialIcon(name, null)
    }
}
